vmaas_sync/metrics.go

```diff
@@ -96,26 +99,58 @@ func update() {
 }
 
 func updateSystemMetrics() {
-	optOuted, notOptOuted, err := getSystemCounts()
+	counts, err := getSystemCounts(time.Now())
 	if err != nil {
 		utils.Log("err", err.Error()).Error("unable to update system metrics")
+		return
+	}
+
+	for key, count := range counts {
+		labels := strings.Split(key, ":")
+		if len(labels) != 2 {
+			utils.Log("nLabels", len(labels), "key", key).Error("unexpected labels count")
+			return
+		}
+		systemsCnt.WithLabelValues(labels...).Set(float64(count))
 	}
-	systemsCnt.WithLabelValues("opt_out_on").Set(float64(optOuted))
-	systemsCnt.WithLabelValues("opt_out_off").Set(float64(notOptOuted))
 }
 
-func getSystemCounts() (optOuted, notOptOuted int, err error) {
+// Load stored systems counts according to "opt_out" and "last_upload" properties.
+// Result is loaded into the map {"opt_out_on:last1D": 12, "opt_out_off:last1D": 3, ...}.
+func getSystemCounts(refTime time.Time) (map[string]int, error) {
 	systemsQuery := database.Db.Model(&models.SystemPlatform{})
-	err = systemsQuery.Where("opt_out = true").Count(&optOuted).Error
-	if err != nil {
-		return 0, 0, errors.Wrap(err, "unable to get metric opt_outed systems")
+	optOutKV := map[string]bool{"opt_out_on": true, "opt_out_off": false}
+	lastUploadKV := map[string]int{"last1D": 1, "last7D": 7, "last30D": 30, "all": -1}
+	counts := map[string]int{}
+	for optOutK, optOutV := range optOutKV {
+		systemsQueryOptOut := updateSystemsQueryOptOut(systemsQuery, optOutV)
+		for lastUploadK, lastUploadV := range lastUploadKV {
+			systemsQueryOptOutLastUpload := updateSystemsQueryLastUpload(systemsQueryOptOut, refTime, lastUploadV)
+			var nSystems int
+			err := systemsQueryOptOutLastUpload.Count(&nSystems).Error
+			if err != nil {
+				return nil, errors.Wrap(err, "unable to load systems counts: "+
+					fmt.Sprintf("opt_out: %v, last_upload_before_days: %v", optOutV, lastUploadV))
+			}
+			key := fmt.Sprintf("%s:%s", optOutK, lastUploadK)
+			counts[key] = nSystems
+		}
 	}
+	return counts, nil
+}
 
-	err = systemsQuery.Where("opt_out = false").Count(&notOptOuted).Error
-	if err != nil {
-		return 0, 0, errors.Wrap(err, "unable to get not opt_outed systems")
+// Update input systems query with "opt_out = X" constraint.
+func updateSystemsQueryOptOut(systemsQuery *gorm.DB, optOut bool) *gorm.DB {
+	return systemsQuery.Where("opt_out = ?", optOut)
+}
+
+// Update input systems query with "last_upload > T" constraint.
+// Constraint is not added if "lastNDays" argument is negative.
+func updateSystemsQueryLastUpload(systemsQuery *gorm.DB, refTime time.Time, lastNDays int) *gorm.DB {
```

Comment:
Weird naming, maybe use  `withLastUploadFilter` ?

Reply:
I used prefix notation here updateSystemsQuery<column-name>. Similarly in case of opt out. I didn't use "with" to make it shorter.